aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -24,12 +24,39 @@ using namespace at::sparse;
 
 namespace {
 
-  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input) {
-    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true)) {
+  template <bool promotes_integer_to_float = false>
+  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input){
+    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true) && promotes_integer_to_float) {
       return at::empty_like(input, input.options().dtype(c10::get_default_dtype()));
     }
     return at::empty_like(input);
   }
+
+  // --------------------------------------------------------------------
+  // helper function for sanity check of input and output tensors
+  // this functions also inits the output tensor
+  // this helper function is for unary ops
+  // --------------------------------------------------------------------
+  template <bool promotes_integer_to_float = false, bool is_linear_map = false>
+  inline void sparse_unary_op_check_and_init(const SparseTensor& self, SparseTensor& result){
+    TORCH_CHECK(result.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.get_device() == result.get_device(),
+        "Both tensors should be on the same device");
+
+    if (promotes_integer_to_float && c10::isIntegralType(self.scalar_type(), /*includeBool=*/true)){
+      TORCH_CHECK(result.scalar_type() == c10::get_default_dtype(), "the result's dtype must be ", c10::get_default_dtype());
+    } else{
+      TORCH_CHECK(result.scalar_type() == self.scalar_type(), "the expected dtype is ", self.scalar_type(),
+          "found ", result.scalar_type());
+    }
+
+    if (is_same_tensor(result, self) && !is_linear_map){
+      TORCH_CHECK(self.is_coalesced(), "inplace variant on uncoalesced is not supported");
```

Comment:
Let's just tweak the language here slightly and suggest a remedy:

"This operation can only be performed inplace on a coalesced sparse tensor; coalesce the tensor before performing it."